<img align="left" src="https://ithaka-labs.s3.amazonaws.com/static-files/images/tdm/tdmdocs/CC_BY.png"><br />

This notebook is adapted by Zhuo Chen from the notebooks created by [Nathan Kelber](http://nkelber.com), [William Mattingly](https://datascience.si.edu/people/dr-william-mattingly) and [Melanie Walsh](https://melaniewalsh.org) under [Creative Commons CC BY License](https://creativecommons.org/licenses/by/4.0/)<br />
For questions/comments/improvements, email zhuo.chen@ithaka.org or nathan.kelber@ithaka.org.<br />
___

# Pandas 2

**Description:** This notebook describes how to:
* Create a dataframe from a .csv file
* Filter data 
* Update data in a dataframe

This is the second notebook in a series on learning to use Pandas. 

**Use Case:** For Learners (Detailed explanation, not ideal for researchers)

**Difficulty:** Intermediate

**Knowledge Required:** 
* [Pandas 1](./pandas-1.ipynb)
* Python Basics ([Start Python Basics I](./python-basics-1.ipynb))

**Knowledge Recommended:** 
* [Python Intermediate 2](./python-intermediate-2.ipynb)
* [Python Intermediate 4](./python-intermediate-4.ipynb)

**Completion Time:** 90 minutes

**Data Format:** CSV (.csv)

**Libraries Used:** Pandas

**Research Pipeline:** None
___


In [ ]:
### Download the sample file for this Lesson
import urllib.request
url = 'https://ithaka-labs.s3.amazonaws.com/static-files/images/tdm/tdmdocs/failed_bank_since_2000.csv'
urllib.request.urlretrieve(url, './data/' + url.rsplit('/', 1)[-1])
print('Samples files retrieved.')

## Create a dataframe from a .csv file

In [Pandas 1](./pandas-1.ipynb), we learned how to create a dataframe by passing a **dictionary** to the `DataFrame` method. We basically entered the data ourselves. 

In real life however, when working with DataFrames, we are usually using a dataset that has been compiled by someone else. Often the data will be in the form of a CSV or Excel file. 

We can convert the data in a .csv file to a Pandas DataFrame using the `.read_csv()` method. We pass in the location of the .csv file.

Use the `**File > Open**` menu above to navigate to the `failed_bank_since_2000.csv` in the `/data` folder. Preview its structure before we load it into a dataframe.

In [1]:
# Import pandas library, `as pd` allows us to shorten typing `pandas` to `pd` when we call pandas
import pandas as pd

In [2]:
# Create a DataFrame `df` from a CSV file using the .read_csv() method
df = pd.read_csv('data/failed_bank_since_2000.csv') # pass in the location of the file
df

Bank Name               City State   Cert  \
0                    Almena State Bank             Almena    KS  15426   
1           First City Bank of Florida  Fort Walton Beach    FL  16748   
2                 The First State Bank      Barboursville    WV  14361   
3                   Ericson State Bank            Ericson    NE  18265   
4     City National Bank of New Jersey             Newark    NJ  21111   
..                                 ...                ...   ...    ...   
558                 Superior Bank, FSB           Hinsdale    IL  32646   
559                Malta National Bank              Malta    OH   6629   
560    First Alliance Bank & Trust Co.         Manchester    NH  34264   
561  National State Bank of Metropolis         Metropolis    IL   3815   
562                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0                            Equity Bank    23-Oct-20  10538  
1              United Fidelity Bank, fsb    16-Oct-20  10537  
2                         MVB Bank, Inc.     3-Apr-20  10536  
3             Farmers and Merchants Bank    14-Feb-20  10535  
4                        Industrial Bank     1-Nov-19  10534  
..                                   ...          ...    ...  
558                Superior Federal, FSB    27-Jul-01   6004  
559                    North Valley Bank     3-May-01   4648  
560  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561              Banterra Bank of Marion    14-Dec-00   4646  
562                   Bank of the Orient    13-Oct-00   4645  

[563 rows x 7 columns]

By default, Pandas displays the first five rows and the last five rows of the dataframe. You can change the display setting using the `.set_option()` method.

In [3]:
# Change the display setting
pd.set_option('display.min_rows', 20) # set the minimum number of rows to display to 30

The display setting is global throughout the notebook. Therefore, any dataframe in the current notebook will have this setting in effect.

In [4]:
# Print out the dataframe after the change in the display setting
df

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
0    15426                          Equity Bank    23-Oct-20  10538  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
..     ...                                  ...          ...    ...  
553  22314                                  NaN     7-Feb-02   4653  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[563 rows x 7 columns]

Now, you see that Pandas displays the first 15 rows and the last 15 rows of the dataframe.

By convention, a dataframe variable is called `df` but we could give it any valid Python variable name. Here, we follow the convention. 

In [ ]:
# Get some info about the dataframe
df.info()

The `info()` method tells us that there are 563 rows and 7 columns in the dataframe. Almost all columns have 563 non-null values, except the column of `Acquiring Institution`.

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

In the exercises in this notebook, we'll work on a dataset built from Constellate.

We'll use the `constellate` client to automatically retrieve the [metadata](https://constellate.org/docs/key-terms/#metadata) for a [dataset](https://constellate.org/docs/key-terms/#dataset). We can retrieve [metadata](https://constellate.org/docs/key-terms/#metadata) in a [CSV file](https://constellate.org/docs/key-terms/#csv-file) using the `get_metadata` method.


In [ ]:
# download and import constellate library
!pip3 install constellate-client
import constellate

In [ ]:
# Creating a variable `dataset_id` to hold our dataset ID
# The default dataset is Shakespeare Quarterly, 1950-present
# retrieve the metadata
dataset_id = "7e41317e-740f-e86a-4729-20dab492e925"
metadata = constellate.get_metadata(dataset_id)

The metadata is stored in a .csv file. In the following code cell, read in the data using Pandas. Give the dataframe a name other than `df`. Then print out the dataframe to take a look. 

Use a Pandas method to explore the dataframe. How many rows does it have? How many columns does it have? What is the data type of the data in each column?

## Filter dataframe
We have learned how to use `.loc` and `.iloc` to select part of a dataframe in [Pandas 1](./pandas-1.ipynb). We will learn more ways to do data filtering in this section.
### Work with missing values
It is a common case that datasets have missing values. As you may have already noticed, blank cells in a CSV file show up as NaN in a Pandas DataFrame. For example, in our dataset, the `Acquiring Institution` column gives the name when a failed bank was acquired by another institution and is empty otherwise.

In [5]:
# Use isna() to check whether a dataframe has missing values
df.isna()

Bank Name   City  State   Cert  Acquiring Institution  Closing Date  \
0        False  False  False  False                  False         False   
1        False  False  False  False                  False         False   
2        False  False  False  False                  False         False   
3        False  False  False  False                  False         False   
4        False  False  False  False                  False         False   
5        False  False  False  False                  False         False   
6        False  False  False  False                  False         False   
7        False  False  False  False                  False         False   
8        False  False  False  False                  False         False   
9        False  False  False  False                  False         False   
..         ...    ...    ...    ...                    ...           ...   
553      False  False  False  False                   True         False   
554      False  False  False  False                  False         False   
555      False  False  False  False                  False         False   
556      False  False  False  False                  False         False   
557      False  False  False  False                  False         False   
558      False  False  False  False                  False         False   
559      False  False  False  False                  False         False   
560      False  False  False  False                  False         False   
561      False  False  False  False                  False         False   
562      False  False  False  False                  False         False   

      Fund  
0    False  
1    False  
2    False  
3    False  
4    False  
5    False  
6    False  
7    False  
8    False  
9    False  
..     ...  
553  False  
554  False  
555  False  
556  False  
557  False  
558  False  
559  False  
560  False  
561  False  
562  False  

[563 rows x 7 columns]

The `.isna()` method put a mask on the original dataframe. The cells with a non-null value are masked with the boolean value of `False`. The cells with a null value are masked with the boolean value of `True`.

We can also use `.isna()` to check whether a specific column has missing values. 

In [7]:
# Use isna() to check whether a column has missing values
df['Acquiring Institution'].isna()

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
       ...  
553     True
554    False
555    False
556    False
557    False
558    False
559    False
560    False
561    False
562    False
Name: Acquiring Institution, Length: 563, dtype: bool

If you want to exclude the rows and columns with missing values from your data analysis, you can use the `.dropna()` method to do that.

By default, the `.dropna()` method drops the rows with at least one missing value. 

In [8]:
# Use .dropna() to remove all rows with at least one missing value
df.dropna() # no argument passed in

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
552                            Net 1st National Bank         Boca Raton    FL   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
0    15426                          Equity Bank    23-Oct-20  10538  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
..     ...                                  ...          ...    ...  
552  26652                       Bank Leumi USA     1-Mar-02   4654  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[532 rows x 7 columns]

You can also set the axis parameter to 0 to drop the rows with missing values.

In [9]:
# Use .dropna() to remove all rows with at least one missing value
df.dropna(axis=0) # Set the axis to 0

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
552                            Net 1st National Bank         Boca Raton    FL   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
0    15426                          Equity Bank    23-Oct-20  10538  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
..     ...                                  ...          ...    ...  
552  26652                       Bank Leumi USA     1-Mar-02   4654  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[532 rows x 7 columns]

Or, you can set the axis parameter to 'rows' drop the rows with missing values.

In [10]:
# Use .dropna() to remove all rows with at least one missing value
df.dropna(axis='rows') # Set the axis to 'rows'

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
552                            Net 1st National Bank         Boca Raton    FL   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
0    15426                          Equity Bank    23-Oct-20  10538  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
..     ...                                  ...          ...    ...  
552  26652                       Bank Leumi USA     1-Mar-02   4654  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[532 rows x 7 columns]

If you set the axis parameter to 1, you will drop the columns with missing values. 

In [11]:
# Use .dropna() to remove all columns with at least one missing value
df.dropna(axis=1) # set the axis to 1

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert Closing Date   Fund  
0    15426    23-Oct-20  10538  
1    16748    16-Oct-20  10537  
2    14361     3-Apr-20  10536  
3    18265    14-Feb-20  10535  
4    21111     1-Nov-19  10534  
5    58317    25-Oct-19  10533  
6    58112    25-Oct-19  10532  
7    10716    31-May-19  10531  
8    30570    15-Dec-17  10530  
9    17719    13-Oct-17  10529  
..     ...          ...    ...  
553  22314     7-Feb-02   4653  
554   8966     1-Feb-02   4652  
555  22002    18-Jan-02   4651  
556  24382    11-Jan-02   4650  
557  34248     7-Sep-01   4649  
558  32646    27-Jul-01   6004  
559   6629     3-May-01   4648  
560  34264     2-Feb-01   4647  
561   3815    14-Dec-00   4646  
562  21029    13-Oct-00   4645  

[563 rows x 6 columns]

You can also drop the columns with missing values by setting the axis parameter to 'columns'.

In [12]:
# Use .dropna() to remove all columns with at least one missing value
df.dropna(axis='columns') # set the axis to 'columns'

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert Closing Date   Fund  
0    15426    23-Oct-20  10538  
1    16748    16-Oct-20  10537  
2    14361     3-Apr-20  10536  
3    18265    14-Feb-20  10535  
4    21111     1-Nov-19  10534  
5    58317    25-Oct-19  10533  
6    58112    25-Oct-19  10532  
7    10716    31-May-19  10531  
8    30570    15-Dec-17  10530  
9    17719    13-Oct-17  10529  
..     ...          ...    ...  
553  22314     7-Feb-02   4653  
554   8966     1-Feb-02   4652  
555  22002    18-Jan-02   4651  
556  24382    11-Jan-02   4650  
557  34248     7-Sep-01   4649  
558  32646    27-Jul-01   6004  
559   6629     3-May-01   4648  
560  34264     2-Feb-01   4647  
561   3815    14-Dec-00   4646  
562  21029    13-Oct-00   4645  

[563 rows x 6 columns]

Sometimes we would want to drop a row only if that row has a missing value in a specific column. We can use the subset parameter to specify the column(s) to look for missing values. 

In [13]:
# Specify the columns to look for missing values
df.dropna(subset=['Acquiring Institution'])

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
552                            Net 1st National Bank         Boca Raton    FL   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
0    15426                          Equity Bank    23-Oct-20  10538  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
..     ...                                  ...          ...    ...  
552  26652                       Bank Leumi USA     1-Mar-02   4654  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[532 rows x 7 columns]

Sometimes you may want to exclude a column from your consideration when you decide whether to drop a row. In other words, even if a row has a missing value in that column, you don't want to drop the row. How do you tell the `.drop()` method to ignore that column?

There is no parameter in the `.drop()` method that we can use to ignore a column, but we can come up with a workaround.

In [94]:
# Put the names of the columns to ignore in a list
col_to_ignore = ['Acquiring Institution']

In [97]:
# Get the names of the columns to consider and put them in a list
col_to_consider = [col for col in df.columns if col not in col_to_ignore]

In [99]:
# Exclude the column(s) to ignore from consideration when dropping rows that have missing values
df.dropna(subset=col_to_consider)

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
0    15426                          Equity Bank    23-Oct-20  10538  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
..     ...                                  ...          ...    ...  
553  22314                                  NaN     7-Feb-02   4653  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[563 rows x 7 columns]

Note that the `.dropna()` method only gives you a preview of the change. The rows or columns with null values are not dropped in the original dataframe. To make the change permanent, you could either assign the result to the variable where you store the original dataframe to update it; or you could use the parameter `inplace=True`.

Sometimes, you would want to maintain the rows and columns that have missing values. However, you would want to fill the cells with NaN values with some values which are of the same data type as the other cells in the same column. In this way, when you apply a certain function to a column in a dataframe, you will not run into type error. A common practice to deal with this kind of problem is to use the `.fillna()` method. 

In [14]:
# Fill the missing values
df['Acquiring Institution'].fillna('No Acquirer')

0                              Equity Bank
1                United Fidelity Bank, fsb
2                           MVB Bank, Inc.
3               Farmers and Merchants Bank
4                          Industrial Bank
5                       Buckeye State Bank
6        Kentucky Farmers Bank Corporation
7                       Legend Bank, N. A.
8                       Royal Savings Bank
9                              Conway Bank
                      ...                 
553                            No Acquirer
554         The State Bank & Trust Company
555       The Security State Bank of Pecos
556       Israel Discount Bank of New York
557                     Delta Trust & Bank
558                  Superior Federal, FSB
559                      North Valley Bank
560    Southern New Hampshire Bank & Trust
561                Banterra Bank of Marion
562                     Bank of the Orient
Name: Acquiring Institution, Length: 563, dtype: object

### Drop columns or rows

We have seen how to drop rows or columns with missing values. Sometimes, even if a row or a column does not have a missing value, you still want to drop them because you will not use them in your analysis anyways. In this case, we will use the `.drop()` method to remove those rows or columns.

You can specify which column you want to drop using the 'columns' parameter. 

In [15]:
# Drop a column by setting the columns parameter
df.drop(columns='Fund')

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date  
0    15426                          Equity Bank    23-Oct-20  
1    16748            United Fidelity Bank, fsb    16-Oct-20  
2    14361                       MVB Bank, Inc.     3-Apr-20  
3    18265           Farmers and Merchants Bank    14-Feb-20  
4    21111                      Industrial Bank     1-Nov-19  
5    58317                   Buckeye State Bank    25-Oct-19  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  
7    10716                   Legend Bank, N. A.    31-May-19  
8    30570                   Royal Savings Bank    15-Dec-17  
9    17719                          Conway Bank    13-Oct-17  
..     ...                                  ...          ...  
553  22314                                  NaN     7-Feb-02  
554   8966       The State Bank & Trust Company     1-Feb-02  
555  22002     The Security State Bank of Pecos    18-Jan-02  
556  24382     Israel Discount Bank of New York    11-Jan-02  
557  34248                   Delta Trust & Bank     7-Sep-01  
558  32646                Superior Federal, FSB    27-Jul-01  
559   6629                    North Valley Bank     3-May-01  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01  
561   3815              Banterra Bank of Marion    14-Dec-00  
562  21029                   Bank of the Orient    13-Oct-00  

[563 rows x 6 columns]

You can drop multiple columns at one time. 

In [16]:
# Drop multiple columns by setting the columns parameter
df.drop(columns=['Fund', 'Cert'])

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

                   Acquiring Institution Closing Date  
0                            Equity Bank    23-Oct-20  
1              United Fidelity Bank, fsb    16-Oct-20  
2                         MVB Bank, Inc.     3-Apr-20  
3             Farmers and Merchants Bank    14-Feb-20  
4                        Industrial Bank     1-Nov-19  
5                     Buckeye State Bank    25-Oct-19  
6      Kentucky Farmers Bank Corporation    25-Oct-19  
7                     Legend Bank, N. A.    31-May-19  
8                     Royal Savings Bank    15-Dec-17  
9                            Conway Bank    13-Oct-17  
..                                   ...          ...  
553                                  NaN     7-Feb-02  
554       The State Bank & Trust Company     1-Feb-02  
555     The Security State Bank of Pecos    18-Jan-02  
556     Israel Discount Bank of New York    11-Jan-02  
557                   Delta Trust & Bank     7-Sep-01  
558                Superior Federal, FSB    27-Jul-01  
559                    North Valley Bank     3-May-01  
560  Southern New Hampshire Bank & Trust     2-Feb-01  
561              Banterra Bank of Marion    14-Dec-00  
562                   Bank of the Orient    13-Oct-00  

[563 rows x 5 columns]

Another way to drop a column is to give the label of the column you want to drop and then set the axis parameter to 1.

In [17]:
# Drop a column by setting the axis parameter
df.drop('Fund', axis=1)

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date  
0    15426                          Equity Bank    23-Oct-20  
1    16748            United Fidelity Bank, fsb    16-Oct-20  
2    14361                       MVB Bank, Inc.     3-Apr-20  
3    18265           Farmers and Merchants Bank    14-Feb-20  
4    21111                      Industrial Bank     1-Nov-19  
5    58317                   Buckeye State Bank    25-Oct-19  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  
7    10716                   Legend Bank, N. A.    31-May-19  
8    30570                   Royal Savings Bank    15-Dec-17  
9    17719                          Conway Bank    13-Oct-17  
..     ...                                  ...          ...  
553  22314                                  NaN     7-Feb-02  
554   8966       The State Bank & Trust Company     1-Feb-02  
555  22002     The Security State Bank of Pecos    18-Jan-02  
556  24382     Israel Discount Bank of New York    11-Jan-02  
557  34248                   Delta Trust & Bank     7-Sep-01  
558  32646                Superior Federal, FSB    27-Jul-01  
559   6629                    North Valley Bank     3-May-01  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01  
561   3815              Banterra Bank of Marion    14-Dec-00  
562  21029                   Bank of the Orient    13-Oct-00  

[563 rows x 6 columns]

You can also drop multiple columns by setting the axis parameter. 

In [18]:
# Drop multiple columns by setting the axis parameter
df.drop(['Fund', 'Cert'], axis=1)

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

                   Acquiring Institution Closing Date  
0                            Equity Bank    23-Oct-20  
1              United Fidelity Bank, fsb    16-Oct-20  
2                         MVB Bank, Inc.     3-Apr-20  
3             Farmers and Merchants Bank    14-Feb-20  
4                        Industrial Bank     1-Nov-19  
5                     Buckeye State Bank    25-Oct-19  
6      Kentucky Farmers Bank Corporation    25-Oct-19  
7                     Legend Bank, N. A.    31-May-19  
8                     Royal Savings Bank    15-Dec-17  
9                            Conway Bank    13-Oct-17  
..                                   ...          ...  
553                                  NaN     7-Feb-02  
554       The State Bank & Trust Company     1-Feb-02  
555     The Security State Bank of Pecos    18-Jan-02  
556     Israel Discount Bank of New York    11-Jan-02  
557                   Delta Trust & Bank     7-Sep-01  
558                Superior Federal, FSB    27-Jul-01  
559                    North Valley Bank     3-May-01  
560  Southern New Hampshire Bank & Trust     2-Feb-01  
561              Banterra Bank of Marion    14-Dec-00  
562                   Bank of the Orient    13-Oct-00  

[563 rows x 5 columns]

To drop a row, you can specify which row you want to drop using the 'index' parameter.

In [19]:
# Drop a row by setting the index parameter
df.drop(index=0)

Bank Name               City State  \
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
10                               Fayette County Bank         Saint Elmo    IL   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
10    1802            United Fidelity Bank, fsb    26-May-17  10528  
..     ...                                  ...          ...    ...  
553  22314                                  NaN     7-Feb-02   4653  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[562 rows x 7 columns]

In the next code cell, can you write some code to drop multiple rows from df?

In [20]:
# Drop multiple rows using the index parameter


Another way to drop a row is to give the label of the row you want to drop and then set the axis parameter to 0.

In [58]:
# Drop a row by setting the axis parameter
df.drop(0, axis=0)

Bank Name               City State  \
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
10                               Fayette County Bank         Saint Elmo    IL   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

      Cert                Acquiring Institution Closing Date   Fund  
1    16748            United Fidelity Bank, fsb    16-Oct-20  10537  
2    14361                       MVB Bank, Inc.     3-Apr-20  10536  
3    18265           Farmers and Merchants Bank    14-Feb-20  10535  
4    21111                      Industrial Bank     1-Nov-19  10534  
5    58317                   Buckeye State Bank    25-Oct-19  10533  
6    58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7    10716                   Legend Bank, N. A.    31-May-19  10531  
8    30570                   Royal Savings Bank    15-Dec-17  10530  
9    17719                          Conway Bank    13-Oct-17  10529  
10    1802            United Fidelity Bank, fsb    26-May-17  10528  
..     ...                                  ...          ...    ...  
553  22314                                  NaN     7-Feb-02   4653  
554   8966       The State Bank & Trust Company     1-Feb-02   4652  
555  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557  34248                   Delta Trust & Bank     7-Sep-01   4649  
558  32646                Superior Federal, FSB    27-Jul-01   6004  
559   6629                    North Valley Bank     3-May-01   4648  
560  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561   3815              Banterra Bank of Marion    14-Dec-00   4646  
562  21029                   Bank of the Orient    13-Oct-00   4645  

[562 rows x 7 columns]

By default, the rows are labeled by their index numbers. You could set the index column to one of the columns of the dataframe. In the next code cell, can you set the index column to the `State` column and then drop all rows with the label 'GA' or 'KS'?

In [22]:
# Drop multiple rows by setting the axis parameter


You might want to drop multiple consecutive rows at one time. The `.drop()` method does not have a parameter for slicing but we can come up with a workaround.

We can use the `.index` property to get the range of indexes for the rows we want to drop and pass them to the `.drop()` method.

In [73]:
# Drop multiple consecutive rows
df.drop(df.index[2:5], axis=0) 

Bank Name               City  \
0                                    Almena State Bank             Almena   
1                           First City Bank of Florida  Fort Walton Beach   
5                                        Resolute Bank             Maumee   
6                                Louisa Community Bank             Louisa   
7                                 The Enloe State Bank             Cooper   
8                  Washington Federal Bank for Savings            Chicago   
9      The Farmers and Merchants State Bank of Argonia            Argonia   
10                                 Fayette County Bank         Saint Elmo   
11   Guaranty Bank, (d/b/a BestBank in Georgia & Mi...          Milwaukee   
12                                      First NBC Bank        New Orleans   
..                                                 ...                ...   
553                                       NextBank, NA            Phoenix   
554                           Oakwood Deposit Bank Co.            Oakwood   
555                              Bank of Sierra Blanca      Sierra Blanca   
556                                  Hamilton Bank, NA              Miami   
557                             Sinclair National Bank           Gravette   
558                                 Superior Bank, FSB           Hinsdale   
559                                Malta National Bank              Malta   
560                    First Alliance Bank & Trust Co.         Manchester   
561                  National State Bank of Metropolis         Metropolis   
562                                   Bank of Honolulu           Honolulu   

    State   Cert                Acquiring Institution Closing Date   Fund  
0      KS  15426                          Equity Bank    23-Oct-20  10538  
1      FL  16748            United Fidelity Bank, fsb    16-Oct-20  10537  
5      OH  58317                   Buckeye State Bank    25-Oct-19  10533  
6      KY  58112    Kentucky Farmers Bank Corporation    25-Oct-19  10532  
7      TX  10716                   Legend Bank, N. A.    31-May-19  10531  
8      IL  30570                   Royal Savings Bank    15-Dec-17  10530  
9      KS  17719                          Conway Bank    13-Oct-17  10529  
10     IL   1802            United Fidelity Bank, fsb    26-May-17  10528  
11     WI  30003  First-Citizens Bank & Trust Company     5-May-17  10527  
12     LA  58302                         Whitney Bank    28-Apr-17  10526  
..    ...    ...                                  ...          ...    ...  
553    AZ  22314                                  NaN     7-Feb-02   4653  
554    OH   8966       The State Bank & Trust Company     1-Feb-02   4652  
555    TX  22002     The Security State Bank of Pecos    18-Jan-02   4651  
556    FL  24382     Israel Discount Bank of New York    11-Jan-02   4650  
557    AR  34248                   Delta Trust & Bank     7-Sep-01   4649  
558    IL  32646                Superior Federal, FSB    27-Jul-01   6004  
559    OH   6629                    North Valley Bank     3-May-01   4648  
560    NH  34264  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
561    IL   3815              Banterra Bank of Marion    14-Dec-00   4646  
562    HI  21029                   Bank of the Orient    13-Oct-00   4645  

[560 rows x 7 columns]

To drop multiple consecutive columns, we can use the `.columns` attribute to get the range of the indexes for the columns we want to drop and then pass it to the `.drop()` method. 

In [72]:
# Drop multiple consecutive columns
df.drop(df.columns[3:5], axis=1)

Bank Name               City State  \
0                                  Almena State Bank             Almena    KS   
1                         First City Bank of Florida  Fort Walton Beach    FL   
2                               The First State Bank      Barboursville    WV   
3                                 Ericson State Bank            Ericson    NE   
4                   City National Bank of New Jersey             Newark    NJ   
5                                      Resolute Bank             Maumee    OH   
6                              Louisa Community Bank             Louisa    KY   
7                               The Enloe State Bank             Cooper    TX   
8                Washington Federal Bank for Savings            Chicago    IL   
9    The Farmers and Merchants State Bank of Argonia            Argonia    KS   
..                                               ...                ...   ...   
553                                     NextBank, NA            Phoenix    AZ   
554                         Oakwood Deposit Bank Co.            Oakwood    OH   
555                            Bank of Sierra Blanca      Sierra Blanca    TX   
556                                Hamilton Bank, NA              Miami    FL   
557                           Sinclair National Bank           Gravette    AR   
558                               Superior Bank, FSB           Hinsdale    IL   
559                              Malta National Bank              Malta    OH   
560                  First Alliance Bank & Trust Co.         Manchester    NH   
561                National State Bank of Metropolis         Metropolis    IL   
562                                 Bank of Honolulu           Honolulu    HI   

    Closing Date   Fund  
0      23-Oct-20  10538  
1      16-Oct-20  10537  
2       3-Apr-20  10536  
3      14-Feb-20  10535  
4       1-Nov-19  10534  
5      25-Oct-19  10533  
6      25-Oct-19  10532  
7      31-May-19  10531  
8      15-Dec-17  10530  
9      13-Oct-17  10529  
..           ...    ...  
553     7-Feb-02   4653  
554     1-Feb-02   4652  
555    18-Jan-02   4651  
556    11-Jan-02   4650  
557     7-Sep-01   4649  
558    27-Jul-01   6004  
559     3-May-01   4648  
560     2-Feb-01   4647  
561    14-Dec-00   4646  
562    13-Oct-00   4645  

[563 rows x 5 columns]

Again, if you want the drop to take place, you will need to assign the resulting dataframe after the drop to the variable storing the original dataframe or set the parameter `inplace=True`.

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

When you explore the Shakespeare dataframe, what did you find about the column `doi`? What did you find about the column `placeOfPublication`? Is there any non-null value in them?

In [ ]:
# Drop the columns of doi and the placeofPublication, make the change permanent


### Filter data using conditionals
Conditional selection using `df.loc[]` is a very common method to filter a dataframe. 

You write a filtering condition to filter a target column. The condition then checks, for each cell in the target column, whether it fulfills the condition or not. The results will be returned as a Series of True/False values. The `.loc` indexer then uses this Series to select the rows that have True values. 

Suppose you are interested in the banks that failed since 2000 in the state of Georgia. From the original dataframe, you would like to get all the rows of the failed banks in Georgia. How do you do it?

In [ ]:
# Write a filtering condition
df['State'] == 'GA' # Returns a Series of True/False values for the column 'State'

In [ ]:
# Assign the filtering condition to a variable
filt = (df['State'] == 'GA') # Use parenthesis for better reading

In [ ]:
# Put the Series returned by the filtering condition within the hard brackets of df.loc[]
df.loc[filt]

Out of the rows that fulfill the filtering condition, we can further specify which columns to be returned.

In [ ]:
# Specify a single column to be returned
df.loc[filt, 'Bank Name']

Of course, we can select muliple columns to be returned out of the filtered rows. 

In [ ]:
# Specify multiple columns to be returned
df.loc[filt, ['Bank Name', 'Fund']]

Now suppose you want to get all the failed banks whose name contains the word 'community'.

In [ ]:
# Get all the banks with the word 'community' in their name
filt = (df['Bank Name'].str.contains('Community'))
df.loc[filt, ['Bank Name']]

#### Conjunction of multiple filtering conditions: `&`

Oftentimes, you would want to filter a dataframe based on more complex conditions. For example, suppose you would like to get the banks in GA that were closed between 2008 and 2010. How do you use `df.loc[ ]` to achieve it?

The location of the failed banks is stored in the `State` column. The closing year of the banks is stored in the `Closing Date` column. 

In [ ]:
# Create the first filtering condition restricting the state
filt1 = (df['State'] == 'GA')

How to get the closing year of the banks? Recall what we have learned in [Pandas 1](./pandas-1.ipynb) about creating a new column based on an old one. How do you extract the closing year out of the column `Closing Date`? We can use the `.apply()` method.

In [ ]:
# Create a new column storing the closing year of the banks
df['Closing Year'] = df['Closing Date'].apply(lambda r: r.split('-')[2])
df['Closing Year'] = df['Closing Year'].astype(int)

In [ ]:
# Take a look at the dataframe
df

In [ ]:
# Create the second filtering condition restricting the closing year
filt2 = (df['Closing Year'] > 7) & (df['Closing Year'] < 11)

With the two filtering conditions, we are ready to extract the banks in GA that failed between 2008 and 2010.

In [ ]:
# Use filt1 and filt2 to get the target rows
df.loc[filt1 & filt2]

Note that when we extract rows that fulfill multiple conditions, we use `&` in Pandas, not `and`. If you replace `&` with `and`, you will get an error. This is different than what we have learned about boolean operators in [Python basics 2](./python-basics-2.ipynb). In Python, we use `and`, `or` and `not`. In Pandas, we use `&`, `|` and `~` intead. 

|Pandas Operator|Boolean|Requires|
|---|---|---|
|&|and|All required to `True`|
|\||or|If any are `True`|
|~|not|The opposite|

Although we use different symbols for these boolean operators, the truth table for them stays the same. For a quick review of the truth table, see [Python basics 2](./python-basics-2.ipynb).

#### Disjunction of multiple filtering conditions: `|`
Suppose you would like to take a look at all the failed banks in the state of Georgia or the state of New York. How do you use `df.loc[ ]` to get the target rows?

In [ ]:
# Create the two filtering conditions restricting the state to GA and NY
filt1 = (df['State'] == 'GA')
filt2 = (df['State'] == 'NY')

In [ ]:
# Use filt1 and filt2 to get the target rows
df.loc[filt1|filt2]

If you would like to get the data of the failed banks in the following six states --- Georgia, New York, New Jersey, Florida, California and West Virginia, you will not want to write six filtering conditions and use the vertical bar `|` to connect all of them. That would be too repetitive. In this case, we can use the `.isin()` method to create a filtering condition.

In [ ]:
# Create a list of the states
states = ['GA', 'NY', 'NJ', 'FL', 'CA', 'WV']

In [ ]:
# Create a filtering condition
filt = (df['State'].isin(states))

In [ ]:
# Use filt to find all failed banks in the six states
df.loc[filt]

#### Negation of a certain condition:`~`
Now, suppose you would like to get all the failed banks that were **not** closed in 2008. How do you do it?

In [ ]:
# Create the filtering condition restricting the closing year to non-2008
filt = (~(df['Closing Year'] == 8))

In [ ]:
# Use the filtering condition to get the target rows with specified columns
df.loc[filt, ['Bank Name', 'City']]

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

Let's do some filtering!

From the Shakespeare dataframe, get the title and the creator of the documents published between 2000 **and** 2010.

From the Shakespeare dataframe, get the creator of the documents shorter than 10 pages **or** longer than 50 pages. 

From the Shakespeare dataframe, get the title of the documents whose publisher is **not** Folger Shakespeare Library. 

## Update a dataframe
We can make changes to the data in a dataframe.
### Update headers
We can update the column names of a dataframe.

In [ ]:
# Take a look at the columns
df.columns

In [ ]:
# Access a column using the dot notation
df.City

In [ ]:
# If a column name has a space in it
df.Bank Name

We could replace all the spaces in column names with an `_`. In this way, we can access all the columns using the dot notation.

In [ ]:
# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')
df

You could also change the case of the headers.

In [ ]:
# Change all headers to upper case
df.columns.str.upper()

We have been updating the column names all at one time. However, oftentimes we just want to update specific columns. In this case, we could use the `df.rename()` method and pass in a **dictionary** where the keys are the original column names and the values are the new column names.

In [ ]:
# Change the column name of 'CERT' to 'CERTIFICATE_NUM'
df.rename(columns = {'Cert':'Certificate_Num'})

To change multiple column names, we just pass in a dictionary to `df.rename()` with multiple key:value pairs.

In [ ]:
# Change multiple column names
df.rename(columns = {'Cert':'Certificate_Num', 'Fund':'Financial_Institution_Num'})

In [ ]:
# Make the change permanent
df.rename(columns = {'Cert':'Certificate_Num', 'Fund':'Financial_Institution_Num'}, inplace = True)

### Update rows 
How to update the values in a row? In [Pandas 1](./pandas-1.ipynb), we have learned how to look up values using `.loc` and `.iloc`.

To update a row, we could use `.loc` or `.iloc` to locate it and then assign the new values to that row.

In [ ]:
# Change an entire row
df.loc[0] = ['Almena State Bank', 'Almena', 'KS', 15426, 'Equity Bank', '23-Oct-20', 10000, 20]

You can locate a specific cell in a row and update the value in that cell alone.

In [ ]:
# Change a specific value in a row
df.loc[0, 'Financial_Institution_Num'] = 10001
df

We could change multiple specific values in a row using `.loc[]`. 

In [ ]:
# Change multiple values in a row
df.loc[0, ['Bank_Name', 'Financial_Institution_Num']] = ['Almena Bank', 12000]
df

### Update columns
There are multiple methods we can use to update columns: `.apply()`, `.map()` and `replace()`.

In [ ]:
# Use apply to update the column CLOSING_YEAR
df['Closing_Year'] = df['Closing_Year'].apply(lambda r:r+2000)
df

In [ ]:
# Use .map() to update specific values in a column
df['Bank_Name'].map({'Almena Bank': 'Almena State Bank', 'The First State Bank': 'West Virginia Bank'})

In [ ]:
# Use .replace() to update specific values in a column while maintaining the rest
df['Bank_Name'].replace({'Almena Bank': 'Almena State Bank', 'The First State Bank': 'West Virginia Bank'})

We can also use a filtering condition to locate the target columns and then make changes. 

For example, we can locate all the banks that failed in 2020 and change their closing date to 'Recent'.

In [ ]:
# Make a filtering condition to get the banks that failed in 2020
filt = (df['Closing_Year'] == 2020)

In [ ]:
# Use the filtering condition to locate the columns and update them
df.loc[filt, ['Financial_Institution_Num', 'Closing_Year']] = [1000, 'Recent']
df

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

Make all column names in the Shakespeare dataframe upper case. 

Get all documents whose current title is 'Review Article' and change their title to 'Review'.

Get all documents whose word count exceeds 5000 and change their word count to the string 'Long article'.

___
## Lesson Complete

Congratulations! You have completed *Pandas 2*.

### Start Next Lesson: [Pandas 3 ->](./pandas-3.ipynb)

### Exercise Solutions
Here are a few solutions for exercises in this lesson.

In [ ]:
# Read in the metadata
shake = pd.read_csv(metadata)

In [ ]:
# Set the rows to display to 30
pd.set_option('display.max_rows', 30)

In [ ]:
# Explore the dataframe
shake.info()

In [ ]:
# Drop the columns of doi and the placeofPublication, make the change permanent
shake.drop(columns=['doi', 'placeOfPublication'], inplace=True)

In [ ]:
# get the title and the creator of the documents published between 2000 and 2010
filt = (shake['publicationYear']>1999) & (shake['publicationYear']<2011)
shake.loc[filt, ['title', 'creator']]

In [ ]:
# get the creator of the documents shorter than 10 pages or longer than 50 pages
filt = (shake['pageCount']<10)|(shake['pageCount']>50)
shake.loc[filt, 'creator']

In [ ]:
# get the title of the documents whose publisher is not Folger Shakespeare Library
filt = (shake['publisher']=='Folger Shakespeare Library')
shake.loc[~filt, 'title']

In [ ]:
# Make all column names in the Shakespeare dataframe upper case
shake.columns = shake.columns.str.upper()

In [ ]:
# Get all documents whose current title is 'Review Article' and change their title to 'Review'
shake.loc[shake['TITLE']=='Review Article', 'TITLE'] = 'Review'

In [ ]:
# Get all documents whose word count exceeds 5000 and change their word count to the string 'Long article'
shake.loc[shake['WORDCOUNT']>5000, 'WORDCOUNT'] = 'Long article'